src/crimson/common/gated.h

```diff
@@ -0,0 +1,61 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/future-util.hh>
+
+#include "crimson/common/exception.h"
+#include "crimson/common/log.h"
+#include "include/ceph_assert.h"
+
+namespace crimson::common {
+
+class Gated {
+ public:
+  virtual void print(std::ostream&) const {}
+ protected:
+  template <typename Func>
+  void gated_dispatch(const char* what, Func&& func) {
+    (void) seastar::with_gate(pending_dispatch, std::forward<Func>(func)
+    ).handle_exception([this, what] (std::exception_ptr&& eptr) {
+      if (*eptr.__cxa_exception_type() == typeid(system_shutdown_exception)) {
+	crimson::get_logger(ceph_subsys_ms).debug(
+	    "{}, {} skipped, system shutdown", *this, what);
+	return seastar::now();
+      }
+      crimson::get_logger(ceph_subsys_ms).error(
```

Comment:
Um, I thought, since this is a header file, I'd better not use that logger() method, since this may conflict with other logger() methods in .cc files. Is this right?

Reply:
I was thinking about the `static` method in the `Gated` class, not outside of it.

Reply:
`gated_logger()` :+1: .